cli/src/formatter.rs

```diff
@@ -612,21 +614,36 @@ impl FormatRecorder {
         mut write_data: impl FnMut(&mut dyn Formatter, Range<usize>) -> io::Result<()>,
     ) -> io::Result<()> {
         let mut last_pos = 0;
-        let mut flush_data = |formatter: &mut dyn Formatter, pos| -> io::Result<()> {
-            if last_pos != pos {
+        let mut flush_data = |formatter: &mut dyn Formatter, pos, raw_text| -> io::Result<()> {
+            if pos == last_pos {
+                return Ok(());
+            }
+            if raw_text {
+                let mut plain_text_formatter = PlainTextFormatter::new(formatter.raw());
+                write_data(&mut plain_text_formatter, last_pos..pos)?;
```

Comment:
I think without this support, raw_text won't work with wrapping :(.

The (/ your) comment in `write_indented` says prefix inherits current labels, which arguably is nothing while raw_text is in play? Basically, it doesn't sound to me like we're breaking the existing behaviour.

But either way, I think `indent()` is the only function that needs this? And do you think this should be blocking or is it okay to fix in a follow up?

Reply:
> I think without this support, raw_text won't work with wrapping :(.

Right. However, wrapping wouldn't work correctly if `raw_text()` contained escaped characters. In your use case, I assume you wouldn't want to count `raw_text()` content as visible characters, so it might make sense to put them as `FormatOp::Raw(Vec<u8>)`. This suggests that `raw_text()` might be a bad function name.

> The (/ your) comment in write_indented says prefix inherits current labels, which arguably is nothing while raw_text is in play?

With the current implementation, indent text will be printed by `PlainTextFormatter` if the trailing text is raw. That's the problem I pointed out. Try something like `indent(label("error", "foo"), raw_text("bar\n"))`.

> And do you think this should be blocking or is it okay to fix in a follow up?

The problem of `indent()` isn't a blocker, but I guess we would need a non-trivial rewrite to fix the problem. So it's nice if you can address it within this PR.

Reply:
> However, wrapping wouldn't work correctly if raw_text() contained escaped characters.

Could you elaborate on this? I was going through the code earlier and there was a comment (again, yours :P) saying textwrap handles escape codes, so I just didn't worry about it. My testing now seems to confirm this as well:

```rust
    #[test]
    fn test_wrap() {
        assert_eq!(
            wrap_bytes(
                b"\x1b]8;;http://example.com\x1b\\Example Text\x1b]8;;\x1b\\",
                12
            ),
            vec![b"\x1b]8;;http://example.com\x1b\\Example Text\x1b]8;;\x1b\\".as_ref()]
        );
        assert_eq!(
            wrap_bytes(
                b"\x1b]8;;http://example.com\x1b\\Example Text\x1b]8;;\x1b\\",
                10
            ),
            vec![
                b"\x1b]8;;http://example.com\x1b\\Example".as_ref(),
                b"Text\x1b]8;;\x1b\\".as_ref(),
            ]
        );
    }
```

(I haven't tested if the 2nd example above is broken by graph nodes in log, but I'm satisfied with this level of support from textwrap.)

---

> With the current implementation, indent text will be printed by PlainTextFormatter if the trailing text is raw. That's the problem I pointed out. Try something like indent(label("error", "foo"), raw_text("bar\n")).

I understand that but I'm not sure why you say it's a problem -- behaviour of `indent(label("prefix", "foo"), label("rest", "bar"))` and `indent(label("prefix", "foo"), raw_text(label("rest", "bar")))` seems consistent to me (which is, the formatting (or the lack of) of the actual content affect the prefix as well).

Reply:
> > However, wrapping wouldn't work correctly if raw_text() contained escaped characters.
> 
> Could you elaborate on this? I was going through the code earlier and there was a comment (again, yours :P) saying textwrap handles escape codes,

Oh, I didn't remember `textrap()` could parse ANSI escapes (because the existing code doesn't depend on it.) Then, the remaining issue is escape sequence containing `\n`? If we add some padding function, it won't work, but we don't have such function right now.

Anyway, if the primary use case is to insert (invisible) escape characters, it should be easier to process these characters as metadata rather than content to be reformatted. The hard part is to find a short descriptive name for it. :)

> > With the current implementation, indent text will be printed by PlainTextFormatter if the trailing text is raw. That's the problem I pointed out. Try something like indent(label("error", "foo"), raw_text("bar\n")).
> 
> I understand that but I'm not sure why you say it's a problem -- behaviour of `indent(label("prefix", "foo"), label("rest", "bar"))` and `indent(label("prefix", "foo"), raw_text(label("rest", "bar")))` seems consistent to me (which is, the formatting (or the lack of) of the actual content affect the prefix as well).

The expected out put for `indent(label("prefix", "foo"), raw_text(label("rest", "bar")))` is `<prefix>foo</prefix>bar`. You'll get `foobar` because `foo` is printed through the PlainTextFormatter.

Reply:
Thanks, added `FormatOp::RawEscapeSequence` with a note in templates.md about nonzero display width.